### PR TITLE
Initial support for bit64::integer64

### PR DESCRIPTION
**Pull request by romainfrancois**
Support for `vec_cast` and `vec_type2`

I'm not sure if I should rather implement a new class `vctrs_integer64` instead so that we can benefit from what `new_vctr` does. It was not clear to me from the vignette how to proceed for existing types.

### CODE REVIEW

File: R/bit64.R [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+#' @export
+vec_type2.integer64.default <- function(x, y) stop_incompatible_type(x, y)
+
+#' @method vec_type2.integer64 integer64
+#' @export
+vec_type2.integer64.integer64 <- function(x, y) bit64::integer64()
+
+#' @method vec_type2.integer64 integer
+#' @export
+vec_type2.integer64.integer <- function(x, y) bit64::integer64()
```
Comment (hadley) [inline R/bit64.R]: I think these are slightly easier to read if you put the pairs (i.e. `vec_type2.integer64.integer` and `vec_type2.integer.integer64` next to each other)

File: R/bit64.R [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+#' @export
+vec_type2.integer64.double <- function(x, y) bit64::integer64()
+
+#' @method vec_type2.integer64 logical
+#' @export
+vec_type2.integer64.logical <- function(x, y) bit64::integer64()
+
+#' @method vec_type2.integer64 character
+#' @export
+vec_type2.integer64.character <- function(x, y) character()
```
Comment (hadley) [inline R/bit64.R]: I _think_ you've been too generous with the automatic coercions here, and we should only automatically coerce integer and logical:

screen shot 2018-10-19 at 7 37 13 am

File: R/bit64.R [hunk 3]
```diff
@@ -0,0 +1,126 @@
+#' @export
```
Comment (hadley) [inline R/bit64.R]: Actually this should be in `type-integer64.R` — can you have a look at the other `type-*.R` files and match their sections?

I think you might also want to make a `new_integer64()` constructor (or maybe `new_int64()`)

File: R/bit64.R [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+#' @method vec_cast.integer64 factor
+vec_cast.integer64.factor <- function(x, to) {
+  bit64::as.integer64(x)
+}
+
+
+#' @export
+#' @method vec_cast.character integer64
+vec_cast.character.integer64 <- function(x, to) {
+  bit64::as.character.integer64(x)
```
Comment (hadley) [inline R/bit64.R]: Why isn't this just `as.character(x)`?

**Comment by hadley** [comment]
Closes #119

**Comment by romainfrancois** [comment]
I think it's better now. Wondering if `vec_data` should be generic too given that it's too aggressive to `integer64` objects, i.e. you get the double representation, and then I had to do this: 

```r
vec_arith_integer64_other <- function(op, x, y){
  c(x, y) %<-% vec_recycle_common(x, y)

  op_fun <- getExportedValue("base", op)
  op_fun(x, vec_data(y))
}
```

specifically to remove the `vec_data(x)` call from `op_fun(vec_data(x), vec_data(y))`

**Comment by hadley** [comment]
I'm reasonable certain it doesn't need to be generic, because it's purpose is to return the low-level data. By its very nature, integer64 requires a bunch of overrides.

**Pull request closed without merging**